File: ch9/py/src/retrieval_graph/graph.py
from typing import Literal
from langchain.hub import pull
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.messages import HumanMessage
from langchain_openai import ChatOpenAI
from langgraph.graph import END, START, StateGraph
from pydantic import BaseModel

from retrieval_graph.utils import format_docs, load_chat_model
from retrieval_graph.configuration import Configuration
from shared.retrieval import make_retriever
from langchain_core.runnables import RunnableConfig

from retrieval_graph.state import AgentState


class Schema(BaseModel):
    route: str = Literal['retrieve', 'direct']
    direct_answer: str


async def check_query_type(state: AgentState, *, config: RunnableConfig):
    configuration = Configuration.from_runnable_config(config)
    structured_llm = load_chat_model(
        configuration.query_model).with_structured_output(Schema)
    routing_prompt = ChatPromptTemplate.from_messages([
        ("system", "You are a routing assistant. Your job is to determine if a question needs document retrieval or can be answered directly.\n\nRespond with either:\n'retrieve' - if the question requires retrieving documents\n'direct' - if the question can be answered directly AND your direct answer"),
        ("human", "{query}")
    ])

    formatted_prompt = routing_prompt.invoke({"query": state["query"]})
    response = structured_llm.invoke(formatted_prompt)

    route = response.route

    if route == "retrieve":
        return {"route": "retrieve_documents"}
    else:
        direct_answer = response.direct_answer
        return {"route": END, "messages": [HumanMessage(content=direct_answer)]}


async def route_query(state: AgentState, *, config: RunnableConfig):
    route = state["route"]
    if not route:
        raise ValueError("Route is not set")

    if route == "retrieve_documents":
        return "retrieve_documents"
    else:
        return END


async def retrieve_documents(state: AgentState, *, config: RunnableConfig):
    configuration = Configuration.from_runnable_config(config)
    retriever = make_retriever(configuration)
    response = retriever.invoke(state["query"])
    return {"documents": response}


async def generate_response(state: AgentState, *, config: RunnableConfig):
    configuration = Configuration.from_runnable_config(config)
    context = format_docs(state["documents"])
    prompt_template = pull("rlm/rag-prompt")
    formatted_prompt = prompt_template.invoke(
        {"context": context, "question": state["query"]})
    messages = formatted_prompt.messages + state["messages"]
    response = load_chat_model(configuration.query_model).invoke(messages)
    return {"messages": response}


builder = StateGraph(AgentState, config_schema=Configuration)
builder.add_node("check_query_type", check_query_type)
builder.add_node("retrieve_documents", retrieve_documents)
builder.add_node("generate_response", generate_response)
builder.add_edge(START, "check_query_type")
builder.add_conditional_edges("check_query_type", route_query)
builder.add_edge("retrieve_documents", "generate_response")
builder.add_edge("generate_response", END)

# Compile into a graph object that you can invoke and deploy.
graph = builder.compile()
graph.name = "RetrievalGraph"

File: ch9/js/src/retrieval_graph/graph.ts
import {
  StateGraph,
  START,
  END,
  MessagesAnnotation,
} from '@langchain/langgraph';
import { AgentStateAnnotation } from './state.js';
import { makeRetriever, makeSupabaseRetriever } from '../shared/retrieval.js';
import { ChatOpenAI } from '@langchain/openai';
import { formatDocs } from './utils.js';
import { ChatPromptTemplate } from '@langchain/core/prompts';
import { pull } from 'langchain/hub';
import { AIMessage, BaseMessage, HumanMessage } from '@langchain/core/messages';
import { z } from 'zod';
import { RunnableConfig } from '@langchain/core/runnables';
import { loadChatModel } from '../shared/utils.js';
import {
  AgentConfigurationAnnotation,
  ensureAgentConfiguration,
} from './configuration.js';
async function checkQueryType(
  state: typeof AgentStateAnnotation.State,
  config: RunnableConfig
): Promise<{
  route: 'retrieve' | 'direct';
  messages?: BaseMessage[];
}> {
  //schema for routing
  const schema = z.object({
    route: z.enum(['retrieve', 'direct']),
    directAnswer: z.string().optional(),
  });

  const configuration = ensureAgentConfiguration(config);
  const model = await loadChatModel(configuration.queryModel);

  const routingPrompt = ChatPromptTemplate.fromMessages([
    [
      'system',
      "You are a routing assistant. Your job is to determine if a question needs document retrieval or can be answered directly.\n\nRespond with either:\n'retrieve' - if the question requires retrieving documents\n'direct' - if the question can be answered directly AND your direct answer",
    ],
    ['human', '{query}'],
  ]);

  const formattedPrompt = await routingPrompt.invoke({
    query: state.query,
  });

  const response = await model
    .withStructuredOutput(schema)
    .invoke(formattedPrompt.toString());

  const route = response.route;

  return { route };
}

async function answerQueryDirectly(
  state: typeof AgentStateAnnotation.State,
  config: RunnableConfig
): Promise<typeof AgentStateAnnotation.Update> {
  const configuration = ensureAgentConfiguration(config);
  const model = await loadChatModel(configuration.queryModel);
  const userHumanMessage = new HumanMessage(state.query);

  const response = await model.invoke([userHumanMessage]);
  return { messages: [userHumanMessage, response] };
}

async function routeQuery(
  state: typeof AgentStateAnnotation.State
): Promise<'retrieveDocuments' | 'directAnswer'> {
  const route = state.route;
  if (!route) {
    throw new Error('Route is not set');
  }
  if (route === 'retrieve') {
    return 'retrieveDocuments';
  } else if (route === 'direct') {
    return 'directAnswer';
  } else {
    throw new Error('Invalid route');
  }
}

async function retrieveDocuments(
  state: typeof AgentStateAnnotation.State,
  config: RunnableConfig
): Promise<typeof AgentStateAnnotation.Update> {
  const retriever = await makeRetriever(config);
  const response = await retriever.invoke(state.query, config);
  return { documents: response };
}

async function generateResponse(
  state: typeof AgentStateAnnotation.State,
  config: RunnableConfig
): Promise<typeof AgentStateAnnotation.Update> {
  const context = formatDocs(state.documents);
  const configuration = ensureAgentConfiguration(config);

  const model = await loadChatModel(configuration.queryModel);
  const promptTemplate = await pull<ChatPromptTemplate>('rlm/rag-prompt');

  const formattedPrompt = await promptTemplate.invoke({
    context: context,
    question: state.query,
  });

  const userHumanMessage = new HumanMessage(state.query);

  // Create a human message with the formatted prompt that includes context
  const formattedPromptMessage = new HumanMessage(formattedPrompt.toString());

  const messageHistory = [...state.messages, formattedPromptMessage];

  // Let MessagesAnnotation handle the message history
  const response = await model.invoke(messageHistory);

  // Return both the current query and the AI response to be handled by MessagesAnnotation's reducer
  return { messages: [userHumanMessage, response] };
}

const builder = new StateGraph(
  AgentStateAnnotation,
  AgentConfigurationAnnotation
)
  .addNode('retrieveDocuments', retrieveDocuments)
  .addNode('generateResponse', generateResponse)
  .addNode('checkQueryType', checkQueryType)
  .addNode('directAnswer', answerQueryDirectly)
  .addEdge(START, 'checkQueryType')
  .addConditionalEdges('checkQueryType', routeQuery, [
    'retrieveDocuments',
    'directAnswer',
  ])
  .addEdge('retrieveDocuments', 'generateResponse')
  .addEdge('generateResponse', END)
  .addEdge('directAnswer', END);

export const graph = builder.compile().withConfig({
  runName: 'RetrievalGraph',
});
